src/box/lua/config/applier/box_cfg.lua

```diff
@@ -428,7 +438,172 @@ local function apply(config)
     box.cfg(box_cfg)
 end
 
+local function missing_names_is_empty(missing_names, configdata)
+    return not missing_names[configdata._replicaset_name] and
+           table.equals(missing_names.peers, {})
+end
+
+local function cluster_on_replace(old, new)
+    -- Ignore insert of new replica. Every new replica will have
+    -- name set, as it may join to a replicaset only by bootstrap, which
+    -- means that config will pass instance_name to box.cfg. If user joins
+    -- replica without config module, we consider he understands, what he's
+    -- doing and don't automatically set instance_name.
+    if old == nil then
+        return
+    end
+
+    local is_replica_drop = old[3] == nil and new == nil
+    local is_name_insert = old[3] == nil and new and new[3] ~= nil
+    if is_name_insert or is_replica_drop then
+        box.on_commit(function()
+            worker:put('apply_names_ch', 'on_name_apply', {new[3]})
+        end)
+    end
+end
+
+local function schema_on_replace(old, new)
+    if old == nil and new and new[1] == 'replicaset_name' then
+        box.on_commit(function()
+            worker:put('apply_names_ch', 'on_name_apply', {new[2]})
+        end)
+    end
+end
+
+local function names_box_cfg(cfg)
+    local ok, err = pcall(function()
+        box.cfg(cfg)
+    end)
+    if not ok then
+        log.warn('Failed to apply name in box.cfg: %s.', err)
+    end
+    return ok
+end
+
+local function names_wait_rw()
+    local watch = nil
+    if box.info.ro then
+        watch = box.watch('box.status', function(_, status)
+            if status.is_ro == false then
+                worker:put('apply_names_ch', 'on_rw_names_apply')
+            end
+        end)
+    else
+        -- Immediately set names if instance is already rw.
+        worker:put('apply_names_ch', 'on_rw_names_apply')
+    end
+    return watch
+end
+
+local function no_missing_names_alerts(config)
+    local configdata = config._configdata
+    if config._alerts[configdata._replicaset_name] ~= nil then
+        return false
+    end
+
+    local peers = configdata:peers()
+    for _, name in ipairs(peers) do
+        if config._alerts[name] ~= nil then
+            return false
+        end
+    end
+
+    return true
+end
+
+local function post_apply(config)
+    -- Instance may receive names during recovery process, triggers
+    -- were not set during box_cfg, so we don't have box.cfg.instance/replicaset
+    -- name configured correctly. Let's set them if they have already been
+    -- received.
+    if box.info.replicaset and
+       box.cfg.replicaset_name ~= box.info.replicaset.name then
+        names_box_cfg({replicaset_name = box.info.replicaset.name})
+    end
+
+    if box.info.name ~= box.cfg.instance_name then
+        names_box_cfg({instance_name = box.info.name})
+    end
+
+    local channel_name = 'apply_names_ch'
```

Comment:
> You use apply_names_ch above too.

Now I don't. I'm trying to avoid global variables, not much sense in them here, everything should be encapsulated inside `apply_names`